File: friend.go
package main

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/lib/pq"
)

// GetPendingFriendRequests 获取待处理的好友请求
// @Summary      获取待处理的好友请求
// @Description  获取当前用户收到但尚未处理的好友请求列表
// @Tags         friends
// @Accept       json
// @Produce      json
// @Security     bearerAuth
// @Success      200  {object}  map[string]interface{}  "待处理请求列表"
// @Router       /api/friends/requests/pending [get]
func (h *Handler) GetPendingFriendRequests(c *gin.Context) {
	userID := c.GetInt64("user_id")
	requests, err := h.svc.friendStore.GetPendingRequests(c.Request.Context(), userID)
	if err != nil {
		handleError(c, ErrGetRequestsFailed)
		return
	}
	list := make([]gin.H, len(requests))
	for i, r := range requests {
		list[i] = gin.H{
			"id":          r.ID,
			"sender_id":   r.FromUserID,
			"receiver_id": r.ToUserID,
			"status":      r.Status,
			"ts":          r.CreatedAt,
		}
	}
	c.JSON(http.StatusOK, gin.H{"requests": list})
}

// SendFriendRequest 发送好友请求
// @Summary      发送好友请求
// @Description  向指定用户发送好友请求
// @Tags         friends
// @Accept       json
// @Produce      json
// @Security     bearerAuth
// @Param        targetPublicId  path  string  true  "目标用户公钥 ID"
// @Success      201  {object}  map[string]interface{}  "请求发送成功"
// @Failure      404  {object}  ErrorResponse  "用户不存在"
// @Failure      409  {object}  ErrorResponse  "请求冲突"
// @Router       /api/friends/request/send/{targetPublicId} [post]
func (h *Handler) SendFriendRequest(c *gin.Context) {
	userID := c.GetInt64("user_id")
	publicID := c.GetString("public_id")
	targetPublicID := c.Param("targetPublicId")

	targetUser, err := h.svc.GetUserByPublicID(c.Request.Context(), targetPublicID)
	if err != nil {
		handleError(c, ErrNotFound)
		return
	}

	requestID, autoAccepted, err := h.svc.SendFriendRequest(c.Request.Context(), userID, targetUser.ID, "")
	if err != nil {
		handleError(c, err)
		return
	}

	status := "pending"
	if autoAccepted {
		status = "auto_accepted"
	}

	c.JSON(http.StatusCreated, gin.H{
		"id":                 requestID,
		"status":             status,
		"sender_public_id":   publicID,
		"receiver_public_id": targetPublicID,
	})
}

// GetFriendRequests 获取好友请求列表
// @Summary      获取好友请求列表
// @Description  获取当前用户发送和收到的所有好友请求
// @Tags         friends
// @Accept       json
// @Produce      json
// @Security     bearerAuth
// @Success      200  {object}  map[string]interface{}  "请求列表"
// @Router       /api/friends/requests [get]
func (h *Handler) GetFriendRequests(c *gin.Context) {
	userID := c.GetInt64("user_id")

	sent, err := h.svc.GetSentRequests(c.Request.Context(), userID)
	if err != nil {
		handleError(c, ErrGetRequestsFailed)
		return
	}

	received, err := h.svc.GetReceivedRequests(c.Request.Context(), userID)
	if err != nil {
		handleError(c, ErrGetRequestsFailed)
		return
	}

	formatRequest := func(r *FriendRequest) gin.H {
		return gin.H{
			"id":          r.ID,
			"sender_id":   r.FromUserID,
			"receiver_id": r.ToUserID,
			"status":      r.Status,
			"ts":          r.CreatedAt,
		}
	}

	sentList := make([]gin.H, len(sent))
	for i, r := range sent {
		sentList[i] = formatRequest(r)
	}

	receivedList := make([]gin.H, len(received))
	for i, r := range received {
		receivedList[i] = formatRequest(r)
	}

	c.JSON(http.StatusOK, gin.H{
		"sent":     sentList,
		"received": receivedList,
	})
}

// AcceptFriendRequest 接受好友请求
// @Summary      接受好友请求
// @Description  接受指定的好友请求
// @Tags         friends
// @Accept       json
// @Produce      json
// @Security     bearerAuth
// @Param        requestId  path  int  true  "请求 ID"
// @Success      200  {object}  SuccessResponse  "接受成功"
// @Failure      400  {object}  ErrorResponse  "请求无效"
// @Failure      403  {object}  ErrorResponse  "无权限"
// @Failure      404  {object}  ErrorResponse  "请求不存在"
// @Router       /api/friends/request/{requestId}/accept [post]
func (h *Handler) AcceptFriendRequest(c *gin.Context) {
	requestID, err := strconv.ParseInt(c.Param("requestId"), 10, 64)
	if err != nil {
		handleError(c, ErrInvalidInput)
		return
	}
	userID := c.GetInt64("user_id")

	if err := h.svc.AcceptFriendRequest(c.Request.Context(), requestID, userID); err != nil {
		handleError(c, err)
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Friend request accepted"})
}

// RejectFriendRequest 拒绝好友请求
// @Summary      拒绝好友请求
// @Description  拒绝指定的好友请求
// @Tags         friends
// @Accept       json
// @Produce      json
// @Security     bearerAuth
// @Param        requestId  path  int  true  "请求 ID"
// @Success      200  {object}  SuccessResponse  "拒绝成功"
// @Failure      400  {object}  ErrorResponse  "请求无效"
// @Failure      404  {object}  ErrorResponse  "请求不存在"
// @Router       /api/friends/request/{requestId}/reject [post]
func (h *Handler) RejectFriendRequest(c *gin.Context) {
	requestID, err := strconv.ParseInt(c.Param("requestId"), 10, 64)
	if err != nil {
		handleError(c, ErrInvalidInput)
		return
	}
	userID := c.GetInt64("user_id")

	if err := h.svc.RejectFriendRequest(c.Request.Context(), requestID, userID); err != nil {
		handleError(c, err)
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Friend request rejected"})
}

// CancelFriendRequest 取消好友请求
// @Summary      取消好友请求
// @Description  取消自己发送的好友请求
// @Tags         friends
// @Accept       json
// @Produce      json
// @Security     bearerAuth
// @Param        requestId  path  int  true  "请求 ID"
// @Success      204  {object}  map[string]interface{}  "取消成功"
// @Failure      400  {object}  ErrorResponse  "请求无效"
// @Failure      404  {object}  ErrorResponse  "请求不存在"
// @Router       /api/friends/request/{requestId} [delete]
func (h *Handler) CancelFriendRequest(c *gin.Context) {
	requestID, err := strconv.ParseInt(c.Param("requestId"), 10, 64)
	if err != nil {
		handleError(c, ErrInvalidInput)
		return
	}
	userID := c.GetInt64("user_id")

	if err := h.svc.CancelFriendRequest(c.Request.Context(), requestID, userID); err != nil {
		handleError(c, err)
		return
	}

	c.JSON(http.StatusNoContent, nil)
}

func (s *Service) SendFriendRequest(ctx context.Context, fromUserID, toUserID int64, message string) (requestID int64, autoAccepted bool, err error) {
	if fromUserID == toUserID {
		return 0, false, ErrSelfRequest
	}

	_, err = s.GetUserByID(ctx, toUserID)
	if err != nil {
		return 0, false, err
	}

	tx, err := s.friendStore.DB().BeginTx(ctx, nil)
	if err != nil {
		return 0, false, err
	}
	defer func() {
		if tx != nil {
			tx.Rollback()
		}
	}()

	reverseID, err := s.friendStore.AcceptPendingTx(tx, toUserID, fromUserID)
	if err != nil {
		return 0, false, err
	}
	if reverseID > 0 {
		if err := tx.Commit(); err != nil {
			return 0, false, err
		}
		tx = nil
		goSafe(func() {
			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
			defer cancel()
			s.SendNotificationToUser(ctx, fromUserID, NotifFriendRequest, NotifData{"request_id": fmt.Sprintf("%d", reverseID), "receiver_uid": toUserID, "action": "accepted"})
		})
		return 0, true, nil
	}

	req := &FriendRequest{
		FromUserID: fromUserID,
		ToUserID:   toUserID,
		Message:    message,
		Status:     "pending",
		CreatedAt:  time.Now().UnixMilli(),
		UpdatedAt:  time.Now().UnixMilli(),
	}

	result, err := s.friendStore.Insert(context.Background(), tx, req)
	if err != nil {
		if isUniqueViolation(err) {
			return 0, false, ErrConflict
		}
		return 0, false, err
	}

	if err := tx.Commit(); err != nil {
		return 0, false, err
	}
	tx = nil

	goSafe(func() {
		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()
		s.SendNotificationToUser(ctx, toUserID, NotifFriendRequest, NotifData{"request_id": fmt.Sprintf("%d", result), "sender_uid": fromUserID, "action": "received"})
	})

	return result, false, nil
}

func isUniqueViolation(err error) bool {
	if pqErr, ok := err.(*pq.Error); ok {
		return pqErr.Code == "23505"
	}
	return false
}

func (s *Service) GetSentRequests(ctx context.Context, fromUserID int64) ([]*FriendRequest, error) {
	return s.friendStore.GetSentRequests(ctx, fromUserID)
}

func (s *Service) GetReceivedRequests(ctx context.Context, toUserID int64) ([]*FriendRequest, error) {
	return s.friendStore.GetReceivedRequests(ctx, toUserID)
}

func (s *Service) AcceptFriendRequest(ctx context.Context, requestID, userID int64) error {
	tx, err := s.friendStore.DB().BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	defer func() {
		if tx != nil {
			tx.Rollback()
		}
	}()

	fromUserID, err := s.friendStore.AcceptByIDTx(tx, requestID, userID)
	if err != nil {
		return err
	}
	if fromUserID == 0 {
		return ErrNotFound
	}
	if fromUserID < 0 {
		return ErrForbidden
	}

	if err := tx.Commit(); err != nil {
		return err
	}
	tx = nil

	goSafe(func() {
		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()
		s.SendNotificationToUser(ctx, fromUserID, NotifFriendRequest, NotifData{"request_id": fmt.Sprintf("%d", requestID), "receiver_uid": userID, "action": "accepted"})
	})

	return nil
}

func (s *Service) RejectFriendRequest(ctx context.Context, requestID, userID int64) error {
	tx, err := s.friendStore.DB().BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	defer func() {
		if tx != nil {
			tx.Rollback()
		}
	}()

	affected, err := s.friendStore.RejectTx(tx, requestID, userID)
	if err != nil {
		return err
	}
	if !affected {
		return ErrNotFound
	}

	if err := tx.Commit(); err != nil {
		return err
	}
	tx = nil
	return nil
}

// 要不要把取消等同于拒绝呢，哦baby不要拒绝我~
func (s *Service) CancelFriendRequest(ctx context.Context, requestID, fromUserID int64) error {
	tx, err := s.friendStore.DB().BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	defer func() {
		if tx != nil {
			tx.Rollback()
		}
	}()

	affected, err := s.friendStore.CancelTx(tx, requestID, fromUserID)
	if err != nil {
		return err
	}
	if !affected {
		return ErrNotFound
	}

	if err := tx.Commit(); err != nil {
		return err
	}
	tx = nil
	return nil
}

// 生成私聊会话 ID，格式为 "p_{较小的 publicID}_{较大的 publicID}"
func PrivateConvID(publicID1, publicID2 string) string {
	if publicID1 < publicID2 {
		return fmt.Sprintf("p_%s_%s", publicID1, publicID2)
	}
	return fmt.Sprintf("p_%s_%s", publicID2, publicID1)
}

func IsPrivateConv(convID string) bool {
	return len(convID) > 0 && convID[0] == 'p'
}

func ParsePrivateConv(convID string) (string, string, error) {
	if !IsPrivateConv(convID) {
		return "", "", errors.New("not a private conversation")
	}
	// 按第一个下划线分割，剩余部分再按最后一个下划线分割
	if len(convID) < 3 || convID[0] != 'p' || convID[1] != '_' {
		return "", "", errors.New("invalid private conversation format")
	}

	rest := convID[2:] // 跳过 "p_"
	lastUnderscore := -1
	for i := len(rest) - 1; i >= 0; i-- {
		if rest[i] == '_' {
			lastUnderscore = i
			break
		}
	}

	if lastUnderscore == -1 {
		return "", "", errors.New("invalid private conversation format")
	}

	a := rest[:lastUnderscore]
	b := rest[lastUnderscore+1:]
	return a, b, nil
}

func CanAccessPrivateConv(convID string, publicID string) bool {
	a, b, err := ParsePrivateConv(convID)
	if err != nil {
		return false
	}
	return a == publicID || b == publicID
}

// 这一段说不定可能会丢到shared里面，但是收益不明显
func ExtractConversationType(convID string) string {
	if len(convID) == 0 {
		return ""
	}
	if convID[0] == 'p' {
		return "p"
	}
	if convID[0] == 'g' {
		return "g"
	}
	return ""
}

func GetOtherPublicID(convID string, myPublicID string) string {
	if !IsPrivateConv(convID) {
		return ""
	}
	a, b, err := ParsePrivateConv(convID)
	if err != nil {
		return ""
	}
	if a == myPublicID {
		return b
	}
	if b == myPublicID {
		return a
	}
	return ""
}

File: src/friend.rs
use axum::Json;
use axum::extract::State;
use axum::http::StatusCode;
use axum::response::IntoResponse;
use serde::Serialize;
use serde_json::json;

use crate::error::AppError;
use crate::shared::Service;
use crate::store::FriendRequest;

#[derive(Debug, Serialize)]
pub struct SendFriendRequestResponse {
    pub id: i64,
    pub status: String,
    pub sender_public_id: String,
    pub receiver_public_id: String,
}

#[derive(Debug, Serialize)]
pub struct FriendRequestResponse {
    pub id: i64,
    pub sender_id: i64,
    pub receiver_id: i64,
    pub status: String,
    pub ts: i64,
}

#[derive(Debug, Serialize)]
pub struct GetFriendRequestsResponse {
    pub sent: Vec<FriendRequestResponse>,
    pub received: Vec<FriendRequestResponse>,
}

impl Service {
    pub async fn send_friend_request(
        &self,
        from_user_id: i64,
        to_user_id: i64,
        message: String,
    ) -> Result<(i64, bool), AppError> {
        if from_user_id == to_user_id {
            return Err(AppError::new(StatusCode::CONFLICT, "Cannot request self"));
        }

        self.get_user_by_id(to_user_id).await?;

        let mut tx = self.pool.begin().await?;

        let reverse_id = self
            .friend_store
            .accept_pending_tx(&mut tx, to_user_id, from_user_id)
            .await?;

        if reverse_id > 0 {
            tx.commit().await?;

            let svc = self.clone();
            let reverse_id_str = reverse_id.to_string();
            tokio::spawn(async move {
                let _ = svc
                    .send_friend_accepted_notification(from_user_id, to_user_id, &reverse_id_str)
                    .await;
            });

            return Ok((0, true));
        }

        let now = chrono::Utc::now().timestamp_millis();
        let req = FriendRequest {
            id: 0,
            from_user_id,
            to_user_id,
            message,
            status: "pending".to_string(),
            created_at: now,
            updated_at: now,
        };

        let result = self
            .friend_store
            .insert_tx(&mut tx, &req)
            .await
            .map_err(|e| {
                if is_unique_violation(&e) {
                    AppError::new(StatusCode::CONFLICT, "Friend request already exists")
                } else {
                    AppError::new(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string())
                }
            })?;

        tx.commit().await?;

        let svc = self.clone();
        let result_str = result.to_string();
        tokio::spawn(async move {
            let _ = svc
                .send_friend_notification(to_user_id, from_user_id, &result_str)
                .await;
        });

        Ok((result, false))
    }

    pub async fn get_sent_requests(
        &self,
        from_user_id: i64,
    ) -> Result<Vec<FriendRequest>, AppError> {
        self.friend_store
            .get_sent_requests(from_user_id)
            .await
            .map_err(|e| AppError::new(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()))
    }

    pub async fn get_received_requests(
        &self,
        to_user_id: i64,
    ) -> Result<Vec<FriendRequest>, AppError> {
        self.friend_store
            .get_received_requests(to_user_id)
            .await
            .map_err(|e| AppError::new(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()))
    }

    pub async fn accept_friend_request(
        &self,
        request_id: i64,
        user_id: i64,
    ) -> Result<(), AppError> {
        let mut tx = self.pool.begin().await?;

        let from_user_id = self
            .friend_store
            .accept_by_id_tx(&mut tx, request_id, user_id)
            .await?;

        if from_user_id == 0 {
            return Err(AppError::not_found("Friend request not found"));
        }
        if from_user_id < 0 {
            return Err(AppError::forbidden());
        }

        tx.commit().await?;

        let svc = self.clone();
        let request_id_str = request_id.to_string();
        tokio::spawn(async move {
            let _ = svc
                .send_friend_accepted_notification(from_user_id, user_id, &request_id_str)
                .await;
        });

        Ok(())
    }

    pub async fn reject_friend_request(
        &self,
        request_id: i64,
        user_id: i64,
    ) -> Result<(), AppError> {
        let mut tx = self.pool.begin().await?;

        let affected = self
            .friend_store
            .reject_tx(&mut tx, request_id, user_id)
            .await?;

        if !affected {
            return Err(AppError::not_found("Friend request not found"));
        }

        tx.commit().await?;

        Ok(())
    }

    pub async fn cancel_friend_request(
        &self,
        request_id: i64,
        from_user_id: i64,
    ) -> Result<(), AppError> {
        let mut tx = self.pool.begin().await?;

        let affected = self
            .friend_store
            .cancel_tx(&mut tx, request_id, from_user_id)
            .await?;

        if !affected {
            return Err(AppError::not_found("Friend request not found"));
        }

        tx.commit().await?;

        Ok(())
    }

    pub async fn block_user(&self, conv_id: &str, target_user_id: i64) -> Result<(), AppError> {
        self.conv_part_store
            .set_blocked(conv_id, target_user_id, true)
            .await
            .map_err(|e| AppError::new(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()))
    }

    pub async fn unblock_user(&self, conv_id: &str, target_user_id: i64) -> Result<(), AppError> {
        self.conv_part_store
            .set_blocked(conv_id, target_user_id, false)
            .await
            .map_err(|e| AppError::new(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()))
    }
}

fn is_unique_violation(err: &sqlx::Error) -> bool {
    if let sqlx::Error::Database(db_err) = err
        && let Some(code) = db_err.code()
    {
        return code == "23505";
    }
    false
}

pub async fn send_friend_request(
    State(state): State<crate::AppState>,
    axum::extract::Extension(user_id): axum::extract::Extension<i64>,
    axum::extract::Extension(public_id): axum::extract::Extension<String>,
    axum::extract::Path(target_public_id): axum::extract::Path<String>,
) -> Result<impl IntoResponse + use<>, AppError> {
    let target_user = state.svc.get_user_by_public_id(&target_public_id).await?;

    let (request_id, auto_accepted) = state
        .svc
        .send_friend_request(user_id, target_user.id, "".to_string())
        .await?;

    let status = if auto_accepted {
        "auto_accepted".to_string()
    } else {
        "pending".to_string()
    };

    Ok((
        StatusCode::CREATED,
        Json(SendFriendRequestResponse {
            id: request_id,
            status,
            sender_public_id: public_id,
            receiver_public_id: target_public_id,
        }),
    ))
}

pub async fn get_friend_requests(
    State(state): State<crate::AppState>,
    axum::extract::Extension(user_id): axum::extract::Extension<i64>,
) -> Result<impl IntoResponse + use<>, AppError> {
    let sent = state.svc.get_sent_requests(user_id).await?;
    let received = state.svc.get_received_requests(user_id).await?;

    let format_request = |r: &FriendRequest| FriendRequestResponse {
        id: r.id,
        sender_id: r.from_user_id,
        receiver_id: r.to_user_id,
        status: r.status.clone(),
        ts: r.created_at,
    };

    let sent_list: Vec<FriendRequestResponse> = sent.iter().map(format_request).collect();
    let received_list: Vec<FriendRequestResponse> = received.iter().map(format_request).collect();

    Ok(Json(GetFriendRequestsResponse {
        sent: sent_list,
        received: received_list,
    }))
}

pub async fn accept_friend_request(
    State(state): State<crate::AppState>,
    axum::extract::Extension(user_id): axum::extract::Extension<i64>,
    axum::extract::Path(request_id): axum::extract::Path<String>,
) -> Result<impl IntoResponse + use<>, AppError> {
    let request_id: i64 = request_id
        .parse()
        .map_err(|_| AppError::bad_request("invalid request_id"))?;
    state.svc.accept_friend_request(request_id, user_id).await?;

    Ok(Json(json!({ "message": "Friend request accepted" })))
}

pub async fn reject_friend_request(
    State(state): State<crate::AppState>,
    axum::extract::Extension(user_id): axum::extract::Extension<i64>,
    axum::extract::Path(request_id): axum::extract::Path<String>,
) -> Result<impl IntoResponse + use<>, AppError> {
    let request_id: i64 = request_id
        .parse()
        .map_err(|_| AppError::bad_request("invalid request_id"))?;
    state.svc.reject_friend_request(request_id, user_id).await?;

    Ok(Json(json!({ "message": "Friend request rejected" })))
}

pub async fn cancel_friend_request(
    State(state): State<crate::AppState>,
    axum::extract::Extension(user_id): axum::extract::Extension<i64>,
    axum::extract::Path(request_id): axum::extract::Path<String>,
) -> Result<impl IntoResponse + use<>, AppError> {
    let request_id: i64 = request_id
        .parse()
        .map_err(|_| AppError::bad_request("invalid request_id"))?;
    state.svc.cancel_friend_request(request_id, user_id).await?;

    Ok(StatusCode::NO_CONTENT)
}

pub async fn block_user(
    State(state): State<crate::AppState>,
    axum::extract::Extension(_user_id): axum::extract::Extension<i64>,
    axum::extract::Extension(public_id): axum::extract::Extension<String>,
    axum::extract::Path(target_public_id): axum::extract::Path<String>,
) -> Result<impl IntoResponse + use<>, AppError> {
    let target_user = state.svc.get_user_by_public_id(&target_public_id).await?;
    let conv_id = crate::shared::private_conv_id(&public_id, &target_public_id);
    state.svc.block_user(&conv_id, target_user.id).await?;
    Ok(Json(json!({ "message": "User blocked successfully" })))
}

pub async fn unblock_user(
    State(state): State<crate::AppState>,
    axum::extract::Extension(_user_id): axum::extract::Extension<i64>,
    axum::extract::Extension(public_id): axum::extract::Extension<String>,
    axum::extract::Path(target_public_id): axum::extract::Path<String>,
) -> Result<impl IntoResponse + use<>, AppError> {
    let target_user = state.svc.get_user_by_public_id(&target_public_id).await?;
    let conv_id = crate::shared::private_conv_id(&public_id, &target_public_id);
    state.svc.unblock_user(&conv_id, target_user.id).await?;
    Ok(Json(json!({ "message": "User unblocked successfully" })))
}

#[derive(Debug, Serialize)]
pub struct GetPendingRequestsResponse {
    pub requests: Vec<FriendRequestResponse>,
}

pub async fn get_pending_requests(
    State(state): State<crate::AppState>,
    axum::extract::Extension(user_id): axum::extract::Extension<i64>,
) -> Result<impl IntoResponse + use<>, AppError> {
    let requests = state
        .svc
        .friend_store
        .get_pending_requests(user_id)
        .await
        .map_err(|e| AppError::new(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()))?;

    let list: Vec<FriendRequestResponse> = requests
        .iter()
        .map(|r| FriendRequestResponse {
            id: r.id,
            sender_id: r.from_user_id,
            receiver_id: r.to_user_id,
            status: r.status.clone(),
            ts: r.created_at,
        })
        .collect();

    Ok(Json(GetPendingRequestsResponse { requests: list }))
}
